File: Iteration2/code/CSE3063GRP12/src/core/repositories/CourseRepository.java
package core.repositories;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import core.database.abstracts.DatabaseManager;
import core.general_providers.AppConstant;
import core.general_providers.InstanceManager;
import core.models.concretes.Course;
import core.models.concretes.CourseEnrollment;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.stream.Stream;

public class CourseRepository {
    private DatabaseManager databaseManager;
    private String path;

    public CourseRepository() {
        databaseManager = InstanceManager.getInstance().getDataBaseInstance();
        path = System.getProperty("user.dir") + AppConstant.getInstance().getBasePath() + "/course/";
    }

    public ArrayList<Course> getCoursesBySemester(int id) throws IOException {
        return readAllJsonFilesInFolder(path + id + "/");
    }

    public void createCourse(Course course) throws IOException {
        databaseManager.write(path + course.getSemester() + "/" + course.getCourseCode() + ".json", course);
    }

    public ArrayList<Course> readAllJsonFilesInFolder(String folderPath) {
        ArrayList<Course> courses = new ArrayList<>();
        Path path = Paths.get(folderPath);
        try (Stream<Path> paths = Files.walk(path)) {
            paths.filter(Files::isRegularFile)
                    .filter(p -> p.toString().endsWith(".json")) // Sadece .json uzantılı dosyaları filtrele
                    .forEach(file -> {
                        Course course = null;
                        try {
                            course = databaseManager.read(file.toString(), Course.class);

                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                        if (course != null) {
                            courses.add(course);
                        }
                    });
        } catch (IOException e) {
            // e.printStackTrace();
        }

        return courses;
    }

    public int getQuota(String courseCode) throws IOException {
        Course course = findCourseByCode(courseCode);

        if (course != null) {
            return course.getQuota();
        } else {
            throw new IllegalArgumentException("Course with code " + courseCode + " not found.");
        }
    }

    public int getCurrentQuota(String courseCode) throws IOException {
        Course course = findCourseByCode(courseCode);

        if (course != null) {
            return course.getCurrentQuota();
        } else {
            throw new IllegalArgumentException("Course with code " + courseCode + " not found.");
        }
    }

    public Course findCourseByCode(String courseCode) throws IOException {
        Path startPath = Paths.get(path);
        try {
            return Files.walk(startPath)
                    .filter(Files::isRegularFile)
                    .filter(path -> path.toString().endsWith(".json"))
                    .filter(path -> path.getFileName().toString().contains(courseCode))
                    .findFirst()
                    .map(path -> {
                        try {
                            return databaseManager.read(path.toString(), Course.class);
                        } catch (Exception e) {
                            System.out.println(e.getMessage());
                            return null;
                        }
                    })
                    .orElse(null);
        } catch (IOException e) {
            System.out.println(e.getMessage());
            return null;
        }
    }

    public ArrayList<Course> findCoursesWithCourseIds(ArrayList<String> ids) {

        ArrayList<Course> matchedCourses = new ArrayList<>();
        Path startPath = Paths.get(path);
        try {
            Files.walk(startPath)
                    .filter(Files::isRegularFile)
                    .filter(path -> path.toString().endsWith(".json"))
                    .filter(path -> ids.stream().anyMatch(id -> path.getFileName().toString().contains(id)))
                    .forEach(path -> {
                        try {
                            Course course = databaseManager.read(path.toString(), Course.class);
                            matchedCourses.add(course);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    });
        } catch (IOException e) {
            // e.printStackTrace();
            System.err.println("Error: " + e.getMessage());
        }

        return matchedCourses;
    }

    public void updateCurrentQuota(CourseEnrollment courseEnrollment) throws IOException {
        if (courseEnrollment == null) {
            System.out.println("CourseEnrollment is null.");
            return;
        }

        //Test
        //System.out.println("updateCurrentQuota is called.");

        ArrayList<Course> approvedCourses = courseEnrollment.getApprovedCourseList();

        if (approvedCourses == null || approvedCourses.isEmpty()) {
            System.out.println("No approved courses found for the student.");
            return;
        }

        for (Course selectedCourse : approvedCourses) {
            String courseCode = selectedCourse.getCourseCode();
            System.out.println("Processing course with code: " + courseCode);

            // Find the corresponding course in the repository
            Course repositoryCourse = findCourseByCode(courseCode);

            if (repositoryCourse != null) {
                // Update the currentQuota for the course in the repository
                int currentQuota = repositoryCourse.getCurrentQuota();
                int updatedQuota = currentQuota + 1;

                //Test
                // System.out.println("Current Quota: " + currentQuota);
                // System.out.println("Updated Quota: " + updatedQuota);

                repositoryCourse.setCurrentQuota(updatedQuota);

                // Save the updated course back to the repository
                databaseManager.write(
                        path + repositoryCourse.getSemester() + "/" + repositoryCourse.getCourseCode() + ".json",
                        repositoryCourse);

                //Test
                //System.out.println("Quota updated successfully for course: " + courseCode);
            } else {
                System.out.println("Course with code " + courseCode + " not found in the repository.");
            }
        }

    }
}

File: Iteration3/code/CSE3063GRP12/src/core/repositories/CourseRepository.py
import json
import os
from pathlib import Path

from core.general_providers.InstanceManager import InstanceManager
from core.general_providers.AppConstant import AppConstant
from core.models.concretes.Course import Course

class CourseRepository:
    def __init__(self):
        self.database_manager = InstanceManager().get_database_instance()
        self.path = os.getcwd() + AppConstant.getInstance().get_base_path() + "course/"
    
    def get_courses_by_semester(self, semester_id):
        folder_path = os.path.join(self.path, str(semester_id))
        return self.read_all_json_files_in_folder(folder_path)   
    
    def create_course(self, course):
        try:
            course_path = os.path.join(self.path, str(course['semester']), f"{course['course_code']}.json")
            self.database_manager.write(course_path, course)
        except Exception as e:
            print(f"An error occurred while creating a course: {e}")

    def read_all_json_files_in_folder(self, folder_path):
        try:
            courses = []
            path = Path(folder_path)
            if path.is_dir():
                for file in path.glob('*.json'):
                    course = self.database_manager.read(file,Course)
                    if course:
                        courses.append(course)
            return courses
        except Exception as e:
            print(f"An error occurred while reading all json files in folder: {e}")
            return None
    
    def findCourseByCode(self, courseCode):
        # walking through the files
        for root, dirs, files in os.walk(self.path):
            for file in files:
                if file.endswith('.json') and courseCode in file:
                    file_path = os.path.join(root, file)
                    try:
                        return self.database_manager.read(file_path, Course)
                    except Exception as e:
                        print(f"Course with {courseCode} is coulnd't found. Error: {e}")
                        return None
        return None
    
    def getQuota(self, courseCode):
        course = self.findCourseByCode(courseCode)
        if course is not None:
            return course.quota
        else:
            raise ValueError(f"Course with code {courseCode} not found.")

    def getCurrentQuota(self, courseCode):
        course = self.findCourseByCode(courseCode)
        if course is not None:
            return course.currentQuota
        else:
            raise ValueError(f"Course with code {courseCode} not found.")
        
    def findCoursesWithCourseIds(self, ids):
        matched_courses = []
        # Walk through the files in the specified path
        for root, dirs, files in os.walk(self.path):
            for file in files:
                # Check if file is a JSON file
                if file.endswith('.json'):
                    # Check if file name contains any of the ids
                    if any(id in file for id in ids):
                        file_path = os.path.join(root, file)
                        try:
                            # Read the course information from the file
                            course = self.database_manager.read(file_path, Course)
                            matched_courses.append(course)
                        except Exception as e:
                            print(e)
        return matched_courses
    
    def updateCurrentQuota(self, courseEnrollment):
        if courseEnrollment is None:
            print("CourseEnrollment is null.")
            return

        approved_courses = courseEnrollment.approved_course_list
        if approved_courses is None or not approved_courses:
            print("No approved courses found for the student.")
            return

        for selected_course in approved_courses:
            course_code = selected_course.courseCode  # Assuming courseCode is an attribute
            print(f"Processing course with code: {course_code}")

            # Find the corresponding course in the repository
            repository_course = self.findCourseByCode(course_code)
            if repository_course is not None:
                # Update the currentQuota for the course in the repository
                repository_course.currentQuota += 1

                # Save the updated course back to the repository
                try:
                    file_path = f"{self.path}/{repository_course.semester}/{repository_course.courseCode}.json"
                    self.database_manager.write(file_path, repository_course)
                    print(f"Quota updated successfully for course: {course_code}")
                except Exception as e:
                    print(f"Error updating quota for {course_code}: {e}")
            else:
                print(f"Course with code {course_code} not found in the repository.")